File: 1306-JumpGameIII/1306-JumpGameIII.cpp
// Last updated: 2/27/2026, 4:40:14 PM
1class Solution {
2public:
3    bool helper(vector<int>& arr, int idx, vector<bool>& visited){
4        if (arr[idx]==0){
5            return true;
6        }
7        if (visited[idx]){
8            return false;
9        }
10        bool goLeft = false;
11        if (idx-arr[idx]>=0){
12            visited[idx] = true;
13            goLeft = helper(arr,idx-arr[idx],visited);
14            visited[idx] = false;
15        }
16        bool goRight = false;
17        if (idx+arr[idx]<arr.size()){
18            visited[idx] = true;
19            goRight = helper(arr,idx+arr[idx],visited);
20            visited[idx] = false;
21        }
22        return goLeft || goRight;
23    }
24    bool canReach(vector<int>& arr, int start) {
25        vector<bool> visited(arr.size(),false);
26        return helper(arr,start,visited);
27    }
28};

File: 1306-JumpGameIII/1306-JumpGameIII.java
// Last updated: 5/17/2026, 7:01:21 AM
1class Solution {
2    public boolean canReach(int[] arr, int start) {
3        boolean[] vis = new boolean[arr.length];
4        return helper(arr,start,vis);
5        
6    }
7    public boolean helper(int[] arr, int st, boolean[] vis){
8        if (st<0 || st>=arr.length || vis[st]){
9            return false;
10        }
11        if (arr[st]==0){
12            return true;
13        }
14
15        vis[st] = true;
16        int back = st-arr[st];
17        int ahead = st+arr[st];
18        boolean res = false;
19
20        
21        
22        res = res || helper(arr,back,vis);
23        res = res || helper(arr,ahead,vis);
24        return res;
25        
26    }
27}